### PR TITLE
164829222 enable users to report articles

### PR DESCRIPTION
**Pull request by ValerieRono**
#### Description
When a user reads an article, he/she should be able to report the article if he/she feels the need to.

#### Type of change
- [ ] Bug fix (nonbreaking change which fixes an issue)
- [x] New feature (nonbreaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

#### How Has This Been Tested?
Please describe the tests that you ran to verify your changes
- [ ] End to end
- [x] Integration

### How can this be manually tested?
1.  Clone the repository 
```https://github.com/andela/ah-centauri-backend.git```

2. cd into project directory and checkout into this branch
```bash 
cd ah-centauri-backend
git checkout feature/164829222-users-report-articles
```

3. Rename the .env.example file into .env and update the parameters with your own keys.

4. Start development server and send the following through postman

The feature has five endpoints, i.e.
 - `POST` report - available to any authenticated and verified user - `/articles/:slug/reports/`
 - `GET` all reports of a particular article - available to any user - `/articles/:slug/reports/`
- `GET` all reports made by a particular user - available to any user - `/articles/reports/get`
 - `GET` specific report - available to any user - `/articles/reports/:id/`
 - `PUT/DELETE` specific report - available to reporter only (must be verified) - `/articles/reports/:id/`

Example request body for POST method:
```
{
    "report" : {
       "report" : "contains vulgar language!",
       "type_of_report" : "Spam"
    }
}
choices for type of report are Spam, Harassment, Rules violation, Plagiarism
```
Example request body for PUT method:
```
{
 "report" : "copied!"
}
```

#### Checklist:
- [x] My code follows the style guide for this project
- [ ] I have linted my code prior to submission
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

#### Pivotal Tracker
- ensure users can post a report to the admins of author's haven
- write tests for said feature

[Delivers #164829222]

### CODE REVIEW

File: authors/apps/core/exceptions.py [hunk 1]
```diff
@@ -52,3 +53,10 @@ def _handle_rating_not_found(exc, context, response):
     response = Response({'errors': 'rating not found'}, status=status.HTTP_404_NOT_FOUND)
```
Comment (ogiste) [inline authors/apps/core/exceptions.py]: I would recommend shifting the error message variables to another file so you can have a single reference for messages when testing as well.

**Review by ogiste — COMMENTED** [review]
Resolve the failing builds

File: authors/apps/articles/urls.py [hunk 2]
```diff
@@ -23,8 +21,13 @@
     path('articles/<slug:slug>/dislike/',
          LikesView.as_view(model=Articles, vote_type=LikeDislike.DISLIKE),
          name='article_dislike'),
-    path('articles/ratings/<int:pk>/', RetrieveUpdateDeleteRatingAPIView.as_view(), name='rating-detail'),
     path('articles/favorites/me/', GetUserFavoritesView.as_view(), name="get_favorites"),
     path('articles/<slug:slug>/favorite/', FavoriteView.as_view(), name="favorite"),
+    path('articles/reports/get',
```
Comment (vino-jasuba) [inline authors/apps/articles/urls.py]: Is there a specific reason why you chose to have the url as `articles/reports/get` ? I'd recommend you use `articles/reports` if the purpose is just listing all reports that have been made just to keep the endpoints consistent

Comment (bl4ck4ndbr0wn) [reply authors/apps/articles/urls.py]: I think the better way to go around is: user `reports/articles/` as a URL

Comment (ValerieRono) [reply authors/apps/articles/urls.py]: done and done,..

File: authors/apps/articles/views.py [hunk 3]
```diff
@@ -430,3 +433,207 @@ def get(self, request):
...
+        Returns
+        --------
+        success message upon deletion of said report
+        errors, if any
+
+        """
+        report = self.get_object(pk)
+        self.check_object_permissions(request, report)
+        report.delete()
+        return Response(status=status.HTTP_204_NO_CONTENT)
```
Comment (ogiste) [inline authors/apps/articles/views.py]: Your delete endpoint also sends a message but the status code is 204, I would recommend changing the status code to 200.

**Review by vino-jasuba — APPROVED** [review]
Looks good to me 👍

**Review by ogiste — APPROVED** [review]

File: authors/apps/articles/views.py (hunk 3 above)
Comment (ValerieRono) [reply authors/apps/articles/views.py]: done

File: authors/apps/core/exceptions.py (hunk 1 above)
Comment (ValerieRono) [reply authors/apps/core/exceptions.py]: done

File: authors/apps/articles/response_messages.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+ERROR_MESSAGES = {
+    "forbidden" : "cannot report own article",
+    "not found" : "no report made by this user",
+    "article" : "article not found",
+    "report" : "report not found",
+    "rating" : "rating not found"
```
Comment (oagutu) [inline authors/apps/articles/response_messages.py]: You can have the `not found messages as one, eg. ```"record_not_found":  "{} not found"```. You can then use `.format()` when accessing the message, passing the appropriate val, eg ```ERROR_MESSAGES["record_not_found"].format("article")``` or sth along these lines

Also add a blank line to the end of this file as per pep8

File: authors/apps/articles/models.py [hunk 5]
```diff
@@ -131,3 +128,32 @@ class Favorite(models.Model):
...
+        (PLAGIARISM, 'Plagiarism'),
+        (SPAM, 'Spam'),
+        (HARASSMENT, 'Harassment'),
+        (RULES_VIOLATION, 'Rules violation'),
+    )
+    author = models.ForeignKey('authentication.User', related_name='reports', on_delete=models.CASCADE, null=False)
+    article =  models.ForeignKey('articles.Articles', related_name='reports', on_delete=models.CASCADE, null=False)
+    report = models.CharField(max_length=140)
+    type_of_report = models.CharField(
+        max_length=20,
```
Comment (oagutu) [inline authors/apps/articles/models.py:149]: No need for `max_length` if you're making use of `choices`

File: authors/apps/articles/views.py [hunk 6]
```diff
@@ -25,6 +26,9 @@ class ArticleNotFound(Exception):
 class RatingNotFound(Exception):
     pass
 
+class ReportNotFound(Exception):
+    pass
+
```
Comment (oagutu) [inline authors/apps/articles/views.py]: Can't the above Exception classes be placed in a separate module?

File: authors/apps/articles/views.py [hunk 7]
```diff
@@ -430,3 +434,204 @@ def get(self, request):
...
+        --------
+        reports made by current user
+        error message if:
+             - any validation errors encountered
+        """
+        reports = ReportArticles.objects.filter(author=request.user)
+        if reports:
+            serializer = ReportsSerializer(reports, many=True)
+            return Response(serializer.data, status=status.HTTP_200_OK)
+        return Response({"errors" : ERROR_MESSAGES['not found']}, status=status.HTTP_400_BAD_REQUEST)
```
Comment (oagutu) [inline authors/apps/articles/views.py]: It shouldn't be a bad request if no reports exist, because there's nothing wrong with the request itself

**Review by oagutu — COMMENTED** [review]
@Val address this among other issues raised:
- report_type should be case sensitive ie. “spam” should not give an invalid choice error message

File: authors/apps/articles/models.py (hunk 5 above)
Comment (ValerieRono) [reply authors/apps/articles/models.py:149]: CharField requires max_length validation

File: authors/apps/articles/views.py (hunk 6 above)
Comment (ValerieRono) [reply authors/apps/articles/views.py]: done

File: authors/apps/articles/views.py (hunk 7 above)
Comment (ValerieRono) [reply authors/apps/articles/views.py]: done

File: authors/apps/articles/response_messages.py (hunk 4 above)
Comment (ValerieRono) [reply authors/apps/articles/response_messages.py]: done

**Pull request merged**